tests/src/test/java/alluxio/master/journal/ufs/RenameFailureJournalTest.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal.ufs;
+
+import static org.junit.Assert.assertTrue;
+
+import alluxio.AlluxioURI;
+import alluxio.LocalAlluxioClusterResource;
+import alluxio.PropertyKey;
+import alluxio.UnderFileSystemFactoryRegistryRule;
+import alluxio.client.file.FileSystem;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.UnderFileSystem.Factory;
+import alluxio.underfs.delegating.DelegatingUnderFileSystem;
+import alluxio.underfs.delegating.DelegatingUnderFileSystemFactory;
+import alluxio.util.CommonUtils;
+
+import com.google.common.io.Files;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Integration test where the rename operation often fails. Makes sure the journal can recover
+ * from this.
+ */
+public class RenameFailureJournalTest {
+  private static final String LOCAL_UFS_PATH = Files.createTempDir().getAbsolutePath();
+
+  private FileSystem mFs;
+
+  // An under file system which fails 90% of its renames.
+  private static final UnderFileSystem UFS =
```

Comment:
Could this behavior be achieved with a spy instead?

Reply:
Yes, but the delegation approach avoids the magic of `Spy`, and is easier to debug (simple to step through). The delegation class mostly writes itself.